### PR TITLE
Add critical sections to avoid USB OTG Errata

### PR DESCRIPTION
**Pull request by jamesmunns**
Lifted from this branch which was tested to prevent spurious stall errors.

https://github.com/embassy-rs/embassy/compare/embassy-rs:e38f101...jamesmunns:3302923

### CODE REVIEW

File: embassy-stm32/src/usb/otg.rs [hunk 1]
```diff
@@ -1208,27 +1214,33 @@ impl<'d, T: Instance> embassy_usb_driver::EndpointIn for Endpoint<'d, T, In> {
...
             });
-        });
 
-        // Write data to FIFO
-        for chunk in buf.chunks(4) {
-            let mut tmp = [0u8; 4];
-            tmp[0..chunk.len()].copy_from_slice(chunk);
-            r.fifo(index).write_value(regs::Fifo(u32::from_ne_bytes(tmp)));
-        }
+            critical_section::with(|_| {
```
Comment (Dirbaio) [inline embassy-stm32/src/usb/otg.rs:1229]: this nested CS isn't doing anything

Comment (jamesmunns) [reply embassy-stm32/src/usb/otg.rs]: @Dirbaio removed!

**Review by Dirbaio — APPROVED** [review]
thank you!

**Comment by mkj** [comment]
@jamesmunns do you remember how to reproduce this? The stm32h7rs doesn't seem to mention that in the errata, I wonder if the critical section can be made conditional there.

**Comment by Dirbaio** [comment]
it was more like "you must access registers in *exactly* the documented order or the peripheral will behave weirdly". Interrupts would cause the order to change, causing endless fun. I don't think this is actually documented as an errata by ST.

**Comment by jamesmunns** [comment]
There was an errata. It shows up in a couple errata docs. If you search for the text "Transmit data FIFO is corrupted when a write sequence to the FIFO is interrupted with accesses to certain OTG_FS registers" and the chip you're interested in, it should show up.

The STM32F405 has it for both USB HS and USB FS:

Screenshot 2025-02-23 at 22 42 21

**Comment by jamesmunns** [comment]
As for how to reproduce, it's very timing based. It would show up every couple days (running 2-3 devices mostly 24/7, talking over postcard-rpc) in my application. You could probably synthetically trigger it, but it would be a pain in the butt.

**Comment by mkj** [comment]
Thanks. I'll see if I can get an answer on the ST forum whether its meant to be fixed.

**Pull request merged**